### PR TITLE
hsm: cleanse HSM token PIN after usage

### PR DESCRIPTION
**Pull request by istepic**

### CODE REVIEW

File: src/mococrw/hsm.h [hunk 1]
```diff
@@ -109,7 +109,7 @@ class HsmEngine : public HSM
     /** Token label used to uniquely identify a token on which objects reside */
     const std::string _tokenLabel;
     /** Pin to access PKCS11 Engine. */
-    const std::string _pin;
+    std::string _pin;
```
Comment (johnfxgalea-e) [inline src/mococrw/hsm.h]: I am wondering why we keep pin in memory and not just cleanse & forget it after hsm construction?

Comment (johnfxgalea-e) [inline src/mococrw/hsm.h]: Why not remove field in class? _ENGINE_ctrl_cmd_string(_engine.get(), "PIN", _pin); is called in constructor.

**Review by johnfxgalea-e — APPROVED** [review]

File: src/mococrw/hsm.h [hunk 2]
```diff
@@ -86,8 +86,8 @@ class HsmEngine : public HSM
 public:
     /**
      * @brief Constructor for an object that can manage keys on HSM using OpenSSL Engine
-     * @note Each HsmEngine object is associated with a specific token and a pin to login to that
-     * token
+     * @note Each HsmEngine object is associated with a specific token
+     * @warning PIN value is not safely cleaned from memory. Make sure you clean it.
```
Comment (johnfxgalea-e) [inline src/mococrw/hsm.h]: How do they clean it - please help the user? Why did we remove the cleanse wrapped function? Couldn't the users use that for convenience?

**Review by johnfxgalea-e — CHANGES_REQUESTED** [review]

File: src/mococrw/hsm.h (hunk 2 above)
Comment (istepic) [reply src/mococrw/hsm.h]: Done

**Review by johnfxgalea-e — APPROVED** [review]
Looks good. Just don't forget to update the changelog

**Review by vkrikun — COMMENTED** [review]
lgtm, but example shall be updates (see comment below).

If we have to deal with PINs, passwords and other sensitive data we should consider creating a "Secret" wrapper class instead of using plain std:. Then we can erase data in destructor.

**Review by vkrikun — CHANGES_REQUESTED** [review]
Pls. update HSM example to erase PIN from memory

**Comment by istepic** [comment]
> If we have to deal with PINs, passwords and other sensitive data we should consider creating a "Secret" wrapper class instead of using plain std:. Then we can erase data in destructor.

Good point. This would break the API and we will have to move to provider interface soon anyways so I recommend we do it like this for now.

**Review by vkrikun — APPROVED** [review]

File: src/mococrw/hsm.h [hunk 3]
```diff
@@ -88,6 +88,8 @@ class HsmEngine : public HSM
      * @brief Constructor for an object that can manage keys on HSM using OpenSSL Engine
      * @note Each HsmEngine object is associated with a specific token and a pin to login to that
      * token
+     * @warning PIN value is not safely cleaned from memory. Make sure you clean it. Check
```
Comment (wusto) [inline src/mococrw/hsm.h:91]:
```suggestion
@warning PIN value is not automatically cleaned from memory in a safe manner after HSM is unlocked. Make sure you clean it after unlocking. Check
```

File: src/mococrw/util.h [hunk 4]
```diff
@@ -78,6 +78,11 @@ void vectorCleanse(std::vector<T> &vec)
     }
```
Comment (wusto) [inline src/mococrw/util.h:78]: Can you please change the line above to also use `openssl::_OPENSSL_cleanse`?

**Pull request merged**